com.playeveryware.eos/Runtime/Core/Config/ProductConfig.cs

```diff
@@ -0,0 +1,172 @@
+            }
+
+            protected PreviousEOSConfig() : base("EpicOnlineServicesConfig.json") { }
+        }
+
+        /// <summary>
+        /// The product ID is a unique GUID labeled "Product ID" in the Epic
+        /// Developer Portal. The name for this value can be set to anything -
+        /// it is used as a label for user interface purposes - and is allowed
+        /// to differ from the label given to it on the Developer Portal.
+        /// </summary>
+        [ConfigField("Product Information", ConfigFieldType.NamedGuid)]
+        public Named<Guid> ProductId;
+
+        /// <summary>
+        /// The set of Clients as defined within the Epic Developer Portal. For
+        /// EOS to function, at least one of these must be set, and the
+        /// platform config needs to indicate which one to use. (If none is
+        /// explicitly indicated, and only one is defined, that one will be
+        /// used).
+        /// </summary>
+        [ConfigField("Client Credentials", ConfigFieldType.ClientCredentials)]
+        public SetOfNamed<EOSClientCredentials> Clients = new("Client");
+
+        /// <summary>
+        /// The set of Sandboxes as defined within the Epic Developer Portal.
+        /// For EOS to function, at least one of these must be set, and it must
+        /// match the deployment indicated by the platform config.
+        /// </summary>
+        [ConfigField("Production Environments", ConfigFieldType.ProductionEnvironments)]
+        public ProductionEnvironments Environments;
+
+        [JsonProperty]
+        private bool _oldConfigImported;
+
+        static ProductConfig()
+        {
+            RegisterFactory(() => new ProductConfig());
+        }
+
+        protected ProductConfig() : base("eos_product_config.json") { }
+
+        private void ImportProductNameAndId(PreviousEOSConfig config)
+        {
+            ProductId ??= new Named<Guid>();
+            ProductId.Name = config.productName;
+            if (!Guid.TryParse(config.productID, out ProductId.Value))
+            {
+                Debug.LogWarning("Could not parse product ID.");
+            }
+        }
+
+        private void ImportClientCredentials(PreviousEOSConfig config)
+        {
+            // Import the old config client stuff
+            Clients.Add(new EOSClientCredentials(config.clientID, config.clientSecret,
+                config.encryptionKey));
+        }
+
+        private void ImportSandboxAndDeployment(PreviousEOSConfig config)
+        {
+            // Import explicitly set sandbox and deployment
+            SandboxId sandboxId = new()
+            {
+                Value = config.sandboxID
+            };
+
+            Deployment oldDeployment = new()
+            {
+                DeploymentId = Guid.Parse(config.deploymentID),
+                SandboxId = sandboxId
+            };
+
+            if (!Environments.AddDeployment(oldDeployment))
```

Comment:
Because these import functions are all called just after the line 
```c#
Environments ??= new();
```
it is not possible for this to be `null`. Since the import functions are `private`, it's not possible for others to improperly call the function at a time when it would be `null`.

Reply:
Sounds believable, ship it.